src/main/java/org/opensearch/security/privileges/PrivilegesEvaluator.java

```diff
@@ -262,7 +262,7 @@ public PrivilegesEvaluatorResponse evaluate(
 
         setUserInfoInThreadContext(user, mappedRoles);
         // Add the security roles for this user so that they can be used for DLS parameter substitution.
-        user.addSecurityRoles(mappedRoles);
+        user.setSecurityRoles(mappedRoles);
```

Comment:
Is it possible to verify that `setSecurityRoles` is only called once per REST Request? IMO it should throw an error if its trying to set a value that's already been set.

Reply:
You can make an atomic reference to the roles and then use SetOnce.